server/handlers.go

```diff
@@ -24,13 +32,169 @@ const (
+			if err != nil {
+				logger.Error("could not generate color hash")
+				return
+			}
+
+			var theme = getTheme(params, logger)
+
+			payload, err = ring.DrawRing(&ring.DrawRingParam{
+				Theme: theme, ColorHash: colorHash, ImageBytes: identityImage.Payload, Height: identityImage.Height, Width: identityImage.Width,
+			})
+
+			if err != nil {
+				logger.Error("failed to draw ring for account identity", zap.Error(err))
+				return
+			}
+		}
+
+		if len(payload) == 0 {
+			logger.Error("empty image")
+			return
+		}
+		mime, err := images.ImageMime(payload)
+		if err != nil {
+			logger.Error("failed to get mime", zap.Error(err))
+		}
+
+		w.Header().Set("Content-Type", mime)
+		w.Header().Set("Cache-Control", "no-store")
+
+		_, err = w.Write(payload)
+		if err != nil {
+			logger.Error("failed to write image", zap.Error(err))
+		}
+	}
+}
+
+func handleContactImages(db *sql.DB, logger *zap.Logger) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		params := r.URL.Query()
+		pks, ok := params["publicKey"]
+		if !ok || len(pks) == 0 {
+			logger.Error("no publicKey")
+			return
+		}
+		imageNames, ok := params["imageName"]
+		if !ok || len(imageNames) == 0 {
+			logger.Error("no imageName")
+			return
+		}
+		colorHash, err := colorhash.GenerateFor(pks[0])
+		if err != nil {
+			logger.Error("could not generate color hash")
+			return
+		}
+
+		var payload []byte
+		err = db.QueryRow(`SELECT payload FROM chat_identity_contacts WHERE contact_id = ? and image_type = ?`, pks[0], imageNames[0]).Scan(&payload)
+		if err != nil {
+			logger.Error("failed to load image.", zap.String("contact id", pks[0]), zap.String("image type", imageNames[0]), zap.Error(err))
+			return
+		}
+
+		if enableRing(params) {
+			var theme = getTheme(params, logger)
+			config, _, err := image.DecodeConfig(bytes.NewReader(payload))
+			if err != nil {
+				logger.Error("failed to decode config.", zap.String("contact id", pks[0]), zap.String("image type", imageNames[0]), zap.Error(err))
+				return
+			}
+
+			payload, err = ring.DrawRing(&ring.DrawRingParam{
+				Theme: theme, ColorHash: colorHash, ImageBytes: payload, Height: config.Height, Width: config.Width,
+			})
+
+			if err != nil {
+				logger.Error("failed to draw ring for contact image.", zap.Error(err))
+				return
+			}
+		}
+
+		if len(payload) == 0 {
+			logger.Error("empty image")
+			return
+		}
+		mime, err := images.ImageMime(payload)
+		if err != nil {
+			logger.Error("failed to get mime", zap.Error(err))
+		}
+
+		w.Header().Set("Content-Type", mime)
+		w.Header().Set("Cache-Control", "no-store")
+
+		_, err = w.Write(payload)
+		if err != nil {
+			logger.Error("failed to write image", zap.Error(err))
+		}
+	}
+}
+
+func enableRing(params url.Values) bool {
+	addRings, ok := params["addRing"]
+	addRing := false
+	if ok && len(addRings) == 1 && addRings[0] == "1" {
+		addRing = true
+	}
+	return addRing
```

Comment:
This could potentially be simplified to:

```suggestion
	return ok && len(addRings) == 1 && addRings[0] == "1"
```